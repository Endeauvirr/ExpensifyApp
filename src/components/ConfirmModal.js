import React from 'react';
import Modal from 'react-modal';

const ConfirmModal = (props) => {
  Modal.setAppElement('body');

  return (
    <Modal
      isOpen={props.isOpen}
      onRequestClose={props.handleCloseModal}
      key={props.id}
      className="modal-confirm"
    >
      <section className="modal__wrapper">
        <header className="modal-header__wrapper">
          <h3 className="modal-header__title">
            Confirm removing expense
          </h3>
        </header>
        <div className="modal-body__wrapper">
          <p>You want to remove <span>{props.description}</span></p>
          <p>Are you sure you want to remove this expense?</p>
        </div>
        <footer className="modal-footer__wrapper">
          <button
            className="modal-footer__cancel-button"
            onClick={props.handleCloseModal}
          >
            Cancel
          </button>
          <button
            className="modal-footer__confirm-button"
            onClick={() => {
              props.handleConfirmModal(props.id);
            }}
          >
            Confirm
          </button>
        </footer>
      </section>
    </Modal>
  );
};

export default ConfirmModal;
