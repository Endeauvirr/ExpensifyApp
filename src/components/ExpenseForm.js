import React from 'react';
import moment from 'moment';

import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount : 0,
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      formValidationError: false
    }
  }
  componentDidMount() {
    if (typeof this.props.expense !== 'undefined') {
      const { description, note, amount, createdAt } = this.props.expense;

      this.setState(() => ({
        description,
        note,
        amount: (amount / 100).toString(),
        createdAt: moment(createdAt)
      }));
    }
  }
  onStringElementChange = (event) => {
    const formFieldName = event.target.name;
    const formFieldValue = event.target.value;

    this.setState(() => ({
      [formFieldName]: formFieldValue
    }))
  }
  onAmountElementChange = (event) => {
    const formFieldName = event.target.name;
    const formFieldValue = event.target.value.replace(',','.');

    if (!formFieldValue || formFieldValue.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        [formFieldName]: formFieldValue
      }))
    }
  }
  onAmountElementFocus = (event) => {
    const value = Number(event.target.value);

    if (value === 0) {
      this.setState(() => ({
        amount: ''
      }))
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  }
  onCalendarFocused = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }))
  }
  onSubmitHandler = (event) => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        formValidationError: true
      }));
    } else {
      this.setState(() => ({
        formValidationError: false
      }));

      const {
        description, amount, createdAt, note
      } = this.state;

      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      });
    }
  }
  render() {
    return (
      <div className="expenses__create-expense--form">
        <form onSubmit={this.onSubmitHandler}>
          {this.state.formValidationError && <p>Please fill description and amount fields</p>}
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={this.state.description}
            onInput={this.onStringElementChange}
          />
          <input
            type="text"
            name="amount"
            placeholder="Amount"
            value={this.state.amount}
            onFocus={this.onAmountElementFocus}
            onChange={this.onAmountElementChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocused}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            name="note"
            placeholder="Place a note for expense (optional)"
            onInput={this.onStringElementChange}
          />
          <button
            type="submit"
          >
            {(this.props.expense) ? 'Edit expense' : 'Create expense'}
          </button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
