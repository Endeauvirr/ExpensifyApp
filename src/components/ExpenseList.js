import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="expenses__list">
    <div className="expenses__list--header">
      <span className="header--label">
        Expense
      </span>
      <span className="header--label">
        Value
      </span>
      <span className="header--label">
        Action
      </span>
    </div>
    <div className="expenses__list--content">
      {
        props.expenses.length === 0 ? (
          <p className="no-expenses-label">No expenses added. Add some more!</p>
        ) : (
          props.expenses.map((expense) => {
            return (
              <ExpenseListItem key={expense.id} {...expense} />
            );
          })
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};


export default connect(mapStateToProps)(ExpenseList);
