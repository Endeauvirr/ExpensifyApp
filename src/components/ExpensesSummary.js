import React from 'react';
import { connect } from 'react-redux';
import getExpensesTotal from '../selectors/expenses-total';
import formatPrice from '../utils/formatPrice';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expensesWord = (expenseCount === 1) ? 'expense' : 'expenses';

  return (
    <div>
      {
        expenseCount === 0 ? (
          <p>You do not have any expenses in this time</p>
        ) : (
          <p>Showing {expenseCount} {expensesWord} with total value of {expensesTotal } </p>
        )
      }
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: formatPrice(getExpensesTotal(visibleExpenses))
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
