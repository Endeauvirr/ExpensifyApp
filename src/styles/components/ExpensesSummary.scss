.expenses__summary {
  p {
    @include primaryTypeface($weight: 300);
    text-align: center;

    span {
      @include primaryTypeface($weight: 700);
    }
  }

  &--wrapper {
    margin-bottom: 40px;
  }
  &--visible {
    font-size: 2em;
    margin-bottom: 0;
  }
  &--overall-info {
    font-size: 0.875em;
    margin-bottom: 0;
    color: $color-prussian-blue;
  }
}