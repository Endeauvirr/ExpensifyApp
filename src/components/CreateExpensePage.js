import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class CreateExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Create a new expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
};

// Przerabiamy ten bezstanowy komponent na jego klasowy odpowiednik by wyciagnąć metodę onSubmit jako props.
// const CreateExpensePage = (props) => (
//   <div>
//     <h1>Create a new expense</h1>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         // props.dispatch(addExpense(expense));  // Zamieniamy to wywołanie, przy pomocy metody z reduxa mapDispathchToProps() , by łatwiej testować funkcję addExpense()
//         props.onSubmit(expense);
//         props.history.push('/');
//       }}
//     />
//   </div>
// );

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
// metoda connect przyjmuje kilka argumentow, undefined na poczatku to argument dla mapStateToProps
