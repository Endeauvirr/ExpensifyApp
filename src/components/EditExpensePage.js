import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
};

// const EditExpensePage = (props) => {
//   console.log(props);

//   return (
//     <div>
//       This is from edit expense page. Im editing the page with ID of {props.match.params.id}

//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.match.params.id, expense));
//           props.history.push('/');
//         }}
//       />
//     </div>
//   );
// };

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
