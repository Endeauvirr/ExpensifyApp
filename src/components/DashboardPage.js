import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';

export const DashboardPage = (props) => {
  return (
    <div>
      <ExpensesSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
};

export default DashboardPage;
