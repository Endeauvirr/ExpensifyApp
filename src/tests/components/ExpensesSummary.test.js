import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should ExpensesSummary component render with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should ExpensesSummary component render with single expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={40000} />);
  expect(wrapper).toMatchSnapshot();
});
