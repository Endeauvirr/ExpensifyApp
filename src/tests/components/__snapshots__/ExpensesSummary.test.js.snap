// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ExpensesSummary component render with no expenses 1`] = `
<div>
  <p>
    You do not have any expenses in this time
  </p>
</div>
`;

exports[`Should ExpensesSummary component render with single expense 1`] = `
<div>
  <p>
    Showing 
    2
     
    expenses
     with total value of 
    40000
     
  </p>
</div>
`;
